plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/inline/InlineChatController.kt

```diff
@@ -0,0 +1,661 @@
+                    recordInlineChatTelemetry(InlineChatUserDecision.DISMISS)
+                    return@runBlocking
+                }
+
+                val isAllEqual = processDiffRows(diff, selectionRange, editor)
+                if (isAllEqual) {
+                    throw Exception("No suggestions from Q; please try a different instruction.")
+                }
+
+                isInProgress.set(false)
+                shouldShowActions.set(true)
+            }
+        }
+    }
+
+    private fun insertNewLineIfNeeded(row: Int, editor: Editor): Int {
+        var newLineInserted = 0
+        while (row > editor.document.lineCount - 1) {
+            insertString(editor, editor.document.textLength, "\n")
+            newLineInserted++
+        }
+        return newLineInserted
+    }
+
+    private fun getLineStartOffset(document: Document, row: Int): Int = ReadAction.compute<Int, Throwable> {
+        document.getLineStartOffset(row)
+    }
+
+    private fun getLineEndOffset(document: Document, row: Int): Int = ReadAction.compute<Int, Throwable> {
+        if (row == document.lineCount - 1) {
+            document.getLineEndOffset(row)
+        } else {
+            document.getLineEndOffset(row) + 1
+        }
+    }
+
+    private fun getLineNumber(document: Document, offset: Int): Int = ReadAction.compute<Int, Throwable> {
+        document.getLineNumber(offset)
+    }
+
+    private fun insertString(editor: Editor, offset: Int, text: String): RangeMarker {
+        var rangeMarker: RangeMarker? = null
+
+        ApplicationManager.getApplication().invokeAndWait {
+            CommandProcessor.getInstance().runUndoTransparentAction {
+                WriteCommandAction.runWriteCommandAction(project) {
+                    editor.document.insertString(offset, text)
+                    val row = editor.document.getLineNumber(offset)
+                    rangeMarker = editor.document.createRangeMarker(offset, getLineEndOffset(editor.document, row))
+                }
+                rangeMarker?.let { marker ->
+                    highlightCodeWithBackgroundColor(editor, marker.startOffset, marker.endOffset, true)
+                }
+            }
+        }
+
+        return rangeMarker!!
+    }
+
+    private fun deleteString(document: Document, start: Int, end: Int) {
+        ApplicationManager.getApplication().invokeAndWait {
+            CommandProcessor.getInstance().runUndoTransparentAction {
+                WriteCommandAction.runWriteCommandAction(project) {
+                    document.deleteString(start, end)
+                }
+            }
+        }
+    }
+
+    private fun highlightString(editor: Editor, start: Int, end: Int, isInsert: Boolean): RangeMarker {
+        var rangeMarker: RangeMarker? = null
+        ApplicationManager.getApplication().invokeAndWait {
+            CommandProcessor.getInstance().runUndoTransparentAction {
+                WriteCommandAction.runWriteCommandAction(project) {
```

Comment:
`ApplicationManager.getApplication().invokeAndWait {}` should give you write lock already so you might not need `WriteCommandAction.runWriteCommandAction(project) {`

Reply:
I think it does need, it will give me an error if I remove this